libs/libpcache/src/faults.cpp

```diff
@@ -0,0 +1,215 @@
+#include <faults/faults.hpp>
+
+using namespace std;
+
+namespace faults {
+
+Fault::Fault(string type) {
+    this->type = type;
+}
+
+Fault::~Fault(){}
+
+
+// Torn seq fault
+ReorderF::ReorderF(string op, vector<int> persist, int occurrence) : Fault(TORN_SEQ) {
+    (this->counter).store(0);
+    this->op = op;
+    this->persist = persist;
+    this->occurrence = occurrence;
+    (this->group_counter).store(0);
+
+}
+
+ReorderF::ReorderF() : Fault(TORN_SEQ) {
+	vector <int> v;
+	(this->counter).store(0);
+    (this->group_counter).store(0);
+    this->op = "";
+	this->persist = v;
+    this->occurrence = 0;
+}
+
+ReorderF::~ReorderF(){}
+
+bool ReorderF::check(string op, vector<int> persist) {
+    if (op != "write") return false;
+
+    bool check = true;
+    for (auto & p : persist) {
+        if (p <= 0) {
+            check = false;
+            break;
+        }
+    }
+    return check;
+}
+
+vector<string> ReorderF::check_with_errors() {
+    vector<string> errors;
+    if (this->op != "write") {
+        errors.push_back("operation must be write");
+    }
+    if (this->occurrence <= 0) {
+        errors.push_back("occurrence must be greater than 0");
+    }
+    for (auto & p : this->persist) {
+        if (p <= 0) {
+            errors.push_back("persist must be greater than 0");
+            break;
+        }
+    }
+    return errors;
+}
+
+
+//Torn op fault
+
+SplitWriteF::SplitWriteF(int occurrence, vector<int> persist, int parts) : Fault(TORN_OP) {
+    (this->counter).store(0);
+    this->occurrence = occurrence;
+    this->persist = persist;
+    this->parts = parts;
+}
+
+SplitWriteF::SplitWriteF(int occurrence, vector<int> persist, vector<int> parts_bytes) : Fault(TORN_OP) {
+    (this->counter).store(0);
+    this->occurrence = occurrence;
+    this->persist = persist;
+    this->parts = -1;
+    this->parts_bytes = parts_bytes;
+}
+
+SplitWriteF::SplitWriteF() : Fault(TORN_OP) {
+	vector <int> v;
+    vector<int> p;
+	(this->counter).store(0);
+    this->occurrence = 0;
+	this->persist = p;
+    this->parts_bytes = v;
+    this->parts = 0;
+}
+
+SplitWriteF::~SplitWriteF(){}
+ 
+bool SplitWriteF::check(vector<int> persist, int parts) {
+    bool check = true;
+    if (parts >= 0) {
+        for (auto & p : persist) {
+            if (p <= 0 || p>parts) {
+                check = false;
+                break;
+            }
+        }
+    } else check = false;
+    return check;
+}
+
+
+bool SplitWriteF::check(vector<int> persist, vector<int> parts_bytes) {
```

Comment:
Can you use more explicit method names? Or shouldn't this be part of a `validate()` function, merged with `SplitWriteF::check_with_errors`?

Reply:
`Check` funtion can indeed be merged with `validate`. Will work on that